import sys
sys.path.insert(1, sys.path[0] + '/../../../library')
from md import process_list
gwas = config['<summary-file>']
gwas = process_list(gwas)
outname = config['--out']
outfolder = config['--outfolder']
geneloc = config['--geneloc']
ldfile = config['--ldfile']
nsamples = config['--nsamples']
genesets = config['--genesets']

base = f"{outfolder}/{outname}"
annot = f"{base}.genes.annot"
raw = expand(base + "{summary}.genes.raw", summary = gwas)
outfile = expand(base + ".{summary}.sets.out", summary = gwas)
genes = expand(base + ".{summary}.genes.out", summary = gwas)
ldfiles = expand(ldfile + ".{ext}", ext=["bim", "bed", "fam"])
bim = f"{ldfile}.bim"


rule all:
    input:
        expand(base + ".{summary}.sets.out", summary = gwas)


rule magma_annot:
    input:
        geneloc, bim
    output:
        annot
    shell:
        "magma --annotate --snp-loc {bim} --gene-loc {geneloc} --out {base}"


rule magma_gene:
    input:
        annot = annot,
        gwas = "{summary}",
        ldfiles = ldfiles
    output:
        raw = base + ".{summary}.genes.raw",
        genes = base + ".{summary}.genes.out"
    shell:
        "magma --bfile {ldfile} --pval {input.gwas} N={nsamples} --gene-annot {input.annot} --out {base}.{wildcards.summary}"


rule magma_geneset:
    input:
        raw = base + ".{summary}.genes.raw",
        genesets = genesets
    output:
        base + ".{summary}.sets.out"
    shell:
        "magma --gene-results {input.raw} --set-annot {input.genesets} --out {base}.{wildcards.summary}"

